### ISSUE TITLE
Faster scaling performance

### ISSUE THREAD

**Issue by wuservices** [sksamuel/scrimage#107]
Just started using scrimage and I was wondering if there's anything in particular I can do to optimize for the fastest performance in either reading or scaling.

I was curious why changing everything to single-threaded addresses scalability as seen in https://github.com/sksamuel/scrimage/commit/0e1444fda19a21d016c7464426c0803e4780086d.

I saw mention that you were going to re-introduce multi-threading in parallelimage. Does that correspond with https://github.com/sksamuel/scrimage/blob/master/scrimage-core/src/main/scala/com/sksamuel/scrimage/ParImage.scala#L540 and is there a reason that this isn't enabled? Would this provide faster scaling?

Thanks for an awesome lib!

**Comment by sksamuel** [comment]
My thinking at the time was this. If you are just scaling one image, you probably don't care that much about speed, as it will be quick anyway. If you are scaling many images, you can easily parallelize by just having each image in a separate executable inside some thread pool executor.

The ParImage is (eventually) going to be like the original where it uses its own thread pool. One of the issues is that the algo needs to know how many segments to create (ideally one per thread right), but from the standard scala ExecutionContext you can't get that info.

**Comment by wuservices** [comment]
Thanks for sharing the thinking. I was trying to improve response time for larger images. On my MacBook Pro Retina, it was taking about 2 seconds to scale down a 20 megapixel image. In contrast, I found that using an imaging web service like imgix to downscale the same image takes about 500 ms including the network time. Of course, they're not scaling images with the JVM, but still as images get bigger, processing time could definitely be improved.

In particular, for my app I'm compositing a bunch of high resolution images to create a web sized preview for something that will eventually be rendered at a different size for print. However, I want to get downscaled versions of all images to build the web preview as quickly as possible as the user is waiting for it.

Could the number of threads just be a param that could be tuned based on the application and environment and is having a setting for `Image.SCALE_THREADS` the only challenge left? Otherwise, is the existing commented out code able to already parallelize correctly?

I could imagine that some applications might want more or less parallelism. In the situation I described, maybe I want to parallelize more or less based on the size of the image, or maybe I want to have fewer threads if the server is serving tons of concurrent users and the threads will be fully utilized. Although I want to reduce some of the overhead of splitting up work, maybe I do want some parallelism to speed things up by 2-4x but no more for a single image.
